agent/connectionuptime/service.go

```diff
@@ -0,0 +1,121 @@
+// pmm-agent
+// Copyright 2019 Percona LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package connectionuptime
+
+import (
+	"sync"
+	"time"
+)
+
+// Service calculates connection uptime between agent and server
+type Service struct {
+	mx sync.Mutex
+
+	connectedStatuses []bool
+
+	windowPeriodSeconds int
+	indexLastStatus     int
+	startTime           time.Time
+	lastStatusTimestamp time.Time
+}
+
+// NewService creates new instance of Service
+func NewService(windowPeriod time.Duration) *Service {
+	return &Service{
+		windowPeriodSeconds: int(windowPeriod.Seconds()),
+		connectedStatuses:   make([]bool, int(windowPeriod.Seconds())),
+	}
+}
+
+// RegisterConnectionStatus adds new connection status
+func (s *Service) RegisterConnectionStatus(timestamp time.Time, connected bool) {
+	s.mx.Lock()
+	defer s.mx.Unlock()
+
+	s.registerConnectionStatus(timestamp, connected)
+}
+
+func (s *Service) registerConnectionStatus(timestamp time.Time, connected bool) {
+	if s.startTime.IsZero() {
+		s.startTime = timestamp
+		s.lastStatusTimestamp = timestamp
+		s.connectedStatuses[0] = connected
+		s.indexLastStatus = 0
+
+		return
+	}
```

Comment:
It looks like some kind of workaround, why timestamp is zero? It should be `current time`, right?

Reply:
Condition `if s.startTime.IsZero() {` checks whether s.startTime is empty or not.

This code will be executed when this method is called first time to initialize some fields